### DISCUSSION TITLE
Problem with testing structs in foundry

### DISCUSSION THREAD

**Discussion by emtothed** [Cyfrin/foundry-full-course-cu#938] [Q&A]
#### Hey everyone

I'm using **foundry** for solidity development and I have some issues with using structs in my tests.

## the problem :
the function below lets users make an offer and saves and pushes the offer in an array of offers in a private mapping. the other function lets the owner of tokens get the offers for their token.in this case, for testing these functions I need to get the offer object in my test.

I tried the library solution and made a library having my offer struct in it and then imported the library in my test file and it worked but in this case, I have to use the library for my whole contract and it feels not okay to me.

isn't there any other way to test my structs in Foundry?
 

### the functions :
```solidity
mapping(uint256 => Offer[]) private tokenIdToOffers;

function makeOffer(uint256 tokenId, uint256 offeredPrice) public {
        if (!_exists(tokenId) || ownerOf(tokenId) == msg.sender) {
            revert wrongTokenIdEnterd();
        }
        if (tokenIdToHighestBid[tokenId] != 0 || tokenIdToIsListed[tokenId]) {
            revert tokenAlreadyInSaleOrAuction();
        }
        Offer memory offer = Offer(msg.sender, offeredPrice);
        tokenIdToOffers[tokenId].push(offer);
    }

function getOffers(
        uint256 tokenId
    ) external view onlyTokenOwner(tokenId) returns (Offer[] memory) {
        return tokenIdToOffers[tokenId];
    }
```
### the test :
```solidity
 function testMakeOffer() public {
        eProp.Offer[] memory offers;
        eProp.mintProp(USER, 44, 65, 1, 1, 0);

        vm.prank(USER2);
        eProp.makeOffer(0, 1 ether);

        vm.prank(USER);
        offers = eProp.getOffers(0);

        assert(offers[0].sender == USER2);
        assert(offers[0].offerdAmount == 1 ether);
    }
```

**Comment by theirrationalone** [comment]
Have you tried this:

```solidity
import "./ExampleContract.sol";

contract NewExample {
    function someFunc() public {
        ExampleContract.MyStruct memory anyStruct = ExampleContract.getMyStruct();
    }
}
```

We don't need any lib. We can access any contract's struct type using `.` member access operator i.e., `ContractName.StructType`.

**Reply by emtothed** [reply]
Hey, sorry I haven't been around for a few days, I tried that but it says: "Identifier not found or not unique" on the struct definition.

**Reply by emtothed** [reply]
I was using contact object instead of the imported contract itself. the problem got solved thanks

**Reply by emtothed** [reply]
I have another problem which I forgot to mention. when the code is getting compiled I get the error below for having too many variables. am I fine and should I just use "--via-ir" or there is something wrong and I should change the code?

#### the error : 
```
Compiler error (/solidity/libyul/backends/evm/AsmCodeGen.cpp:63):Stack too deep. Try compiling with --via-ir (cli) or the equivalent viaIR: true (standard JSON) while enabling the optimizer. Otherwise, try removing local variables. When compiling inline assembly: Variable pos is 3 slot(s) too deep inside the stack. Stack too deep. Try compiling with --via-ir (cli) or the equivalent viaIR: true (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
```

**Reply by theirrationalone** [reply]
You need to optimize your codebase, somehow execution is getting into a nesting hell. 

Btw i know a Guy who is doing & knows tricks to deal with this. He is literally cool.... 

Heyy @usmanfarooq91 !, would you like to help @emtothed !.

**Reply by 0xusmanf** [reply]
@theirrationalone thanks for the call and kind words, @emtothed I have discussed this issue [here](https://github.com/Cyfrin/foundry-full-course-f23/discussions/851#discussioncomment-6992652) in detail on another similar issue, maybe it can help you understand the problem.

**Reply by emtothed** [reply]
Hey @usmanfarooq91 thanks for helping.
I've read the discussion you linked and I found the problematic function, it was my tokenURI function which looks like this : 
```solidity
function tokenURI(
        uint256 tokenId
    ) public view override returns (string memory) {
        string memory length = Strings.toString(tokenIdToSpec[tokenId].length);
        string memory width = Strings.toString(tokenIdToSpec[tokenId].width);
        string memory x = Strings.toString(tokenIdToSpec[tokenId].x);
        string memory y = Strings.toString(tokenIdToSpec[tokenId].y);
        string memory propType = propTypeToString(
            tokenIdToSpec[tokenId].propType
        );
        return
            string(
                abi.encodePacked(
                    _baseURI(),
                    Base64.encode(
                        bytes(
                            abi.encodePacked(
                                '{"name":"',
                                name(),
                                '","description":"EProp is a decentralized property ownership system.","attributes":[{"Length":"',
                                length,
                                '","Width":"',
                                width,
                                '","X":"',
                                x,
                                '","Y":"',
                                y,
                                '","Property type":"',
                                propType,
                                '"}],"image":"',
                                s_imageUri,
                                '"}'
                            )
                        )
                    )
                )
            );
    }
```
**I changed it in the way below and removed the local variables but it's still the same**
```solidity
function tokenURI(
        uint256 tokenId
    ) public view override returns (string memory) {
        return
            string(
                abi.encodePacked(
                    _baseURI(),
                    Base64.encode(
                        bytes(
                            abi.encodePacked(
                                '{"name":"',
                                name(),
                                '","description":"EProp is a decentralized property ownership system.","attributes":[{"Length":"',
                                tokenIdToSpec[tokenId].length.toString(),
                                '","Width":"',
                                tokenIdToSpec[tokenId].width.toString(),
                                '","X":"',
                                tokenIdToSpec[tokenId].x.toString(),
                                '","Y":"',
                                tokenIdToSpec[tokenId].y.toString(),
                                '","Property type":"',
                                propTypeToString(
                                    tokenIdToSpec[tokenId].propType
                                ),
                                '"}],"image":"',
                                s_imageUri,
                                '"}'
                            )
                        )
                    )
                )
            );
    }
```
**but when I totally remove one of the variables it works ok, however, I need all of them in my tokenURI. what can I do in this case? I really appreciate your guidance.**

**Reply by emtothed** [reply]
@theirrationalone thanks a lot for helping 😃 🙌

**Reply by 0xusmanf** [reply]
@emtothed Since I don't have your repo I haven't tried and don't really know if it will work but you can try the following solution.
```solidity
function tokenURI(uint256 tokenId) public view override returns (string memory) {
        string memory encodedData = abi.encodePacked(
            '{"name":"',
            name(),
            '","description":"EProp is a decentralized property ownership system.","attributes":[{"Length":"',
            tokenIdToSpec[tokenId].length.toString(),
            '","Width":"',
            tokenIdToSpec[tokenId].width.toString(),
            '","X":"',
            tokenIdToSpec[tokenId].x.toString(),
            '","Y":"',
            tokenIdToSpec[tokenId].y.toString(),
            '","Property type":"',
            propTypeToString(tokenIdToSpec[tokenId].propType),
            '"}],"image":"',
            s_imageUri,
            '"}'
        );

        return string(abi.encodePacked(_baseURI(), Base64.encode(bytes(encodedData))));
    }
```

**Reply by emtothed** [reply]
@usmanfarooq91 Hey friend, I tried the code and unfortunately it didn't work . I fixed it by merging "x" and "y" into "location" like this :
```solidity
 string memory location = string.concat(
            Strings.toString(tokenIdToSpec[tokenId].x),
            " , ",
            Strings.toString(tokenIdToSpec[tokenId].y)
        );
```
but I don't fully understand what was going on, can you enlighten me about the technical details?

**Comment by GenevaRoc** [comment]
I'm facing similar struct issue.
here is the test for it. 
` function testCanCreateGroup() public registerUsers {
        vm.startPrank(USER1);
        groupSavings.createSavingsGroup(goalAmount, releaseTime);
        // Check if the savings group was created successfully
        // Get the savings groups array
       // SavingsGroup[] memory groups = groupSavings.getSavingsGroups();
        //groupSavings.getSavingsGroups();
        // Verify the details of the last savings group
       //savingsGroups memory group = groups[groups.length - 1];
       GroupSavings.SavingsGroup memory groupsave = groupSavings.getSavingsGroups();

        assertEq(GroupSavings.creator, msg.sender, "Creator is not correct");
        assertEq(GroupSavings.goalAmount, goalAmount, "Goal amount is not correct");
        assertEq(GroupSavings.releaseTime, releaseTime, "Release time is not correct");
        vm.stopPrank();
    }
`


and here is the code I want to test for 
`function createSavingsGroup(uint256 _goalAmount, uint256 _releaseTime) external {
        if (_releaseTime < block.timestamp) {
            revert TalkTrove_ReleaseTimeMustBeInFuture();
        }

        bool isUser = i_tT._checkUserIsRegistered(msg.sender);
        if (!isUser) {
            revert TalkTrove_MustBeRegistered();
        }
        s_theTimeGiven = TIME_STAMP_CAL * _releaseTime;
        // uint256 currentTimeStamp = block.timestamp;

        uint256 newGroupId = savingsGroups.length - 1;
        SavingsGroup storage group = savingsGroups[newGroupId];
        group.creator = msg.sender;
        group.goalAmount = _goalAmount;
        group.releaseTime = s_theTimeGiven;
        savingsGroups.push();
        emit NewSavingsGroup(msg.sender, _goalAmount, _releaseTime);
    }`

this is what the struct looks like 
`struct SavingsGroup {
        address creator;
        uint256 goalAmount;
        uint256 releaseTime;
        mapping(address => uint256) contributions;
        mapping(address => bool) members;
        address[] contributors;
    }`

I can't access the struct in the createSavingsGroup function from my test file. I tried creating a getter function to tackle the issue but it throws an error with the nested mappings

**Reply by theirrationalone** [reply]
Sorry if i guide you wrong... (Your formatting needs to be better) 

Generally We can access a struct as a contract member via member access operator `.`

Look at this example.... 

```javascript
import {MyContract} from "contract/directory/MyContract.sol";

Contract TestContract {
    function play() public {
        MyContract.MyStruct memory _myStruct;
        // now you can access that struct members from memory via again member access operator `.`
    }
}
```

`MyStruct` is a struct that has definition inside `MyContract` contract.
